### PR TITLE
JIT must narrow byte/short/char before storing

### PR DESCRIPTION
**Pull request by BradleyWood**
Issue: #16498

### CODE REVIEW

**Comment by gacholio** [comment]
Will this change also impact Unsafe putObject* ? I'll next be looking at the interpreter to see if they need changing.

**Comment by BradleyWood** [comment]
This only affects putstatic/putfield

**Review by hzongaro — COMMENTED** [review]
I notice that pull request #18898 adjusts the implementation of `withfield` in the JVM interpreter, in addition to `putfield` and `putstatic`.  Although it probably won't be around much longer, may I ask you to make the corresponding change to IL Generation for `withfield` to maintain consistency?

May I also ask you to add some details to the commit comment describing the reason for the change?

**Comment by gacholio** [comment]
I've also modified the pre-OpenJdK MethodHandle implementation in the VM (#18929). I suspect the JIT will need no work here because the old implementation boiled down to bytecodes, so hopefully this case will be fixed by the existing work.

File: runtime/compiler/ilgen/Walker.cpp [hunk 1]
```diff
@@ -6998,6 +6998,42 @@ TR_J9ByteCodeIlGenerator::storeInstance(int32_t cpIndex)
...
+            }
+         else
+            {
+            value = TR::Node::create(TR::i2s, 1, value);
+            value = TR::Node::create(TR::s2i, 1, value);
+            }
+         break;
+      default:
+         break;
+      }
```
[the remark this answers is not in the document]
Comment (BradleyWood) [reply runtime/compiler/ilgen/Walker.cpp:7043]: See the [force-push](https://github.com/eclipse-openj9/openj9/compare/02de32fc17cf5a9e247a97b4a142fb4703aa23d2..0c89f2941b6d23780dd0ebd6157c73042b52e5ed)

**Comment by hzongaro** [comment]
> Then you'll need to do something similar for the Unsafe intrinsics (the VM already did this right for a long time, with the boolean special case added recently).

@BradleyWood , you mentioned offline that the handling of `Unsafe` is already as expected for JIT-compiled code.  Can you confirm that?

File: runtime/compiler/ilgen/J9ByteCodeIlGenerator.hpp [hunk 2]
```diff
@@ -205,6 +205,7 @@ class TR_J9ByteCodeIlGenerator : public TR_IlGenerator, public TR_J9ByteCodeIter
    void         storeFlattenableInstance(int32_t);
    void         storeFlattenableInstanceWithHelper(int32_t);
    void         storeStatic(int32_t);
+   TR::Node*    narrowIntStore(TR::Node *value, TR::SymbolReference *symRef);
```
Comment (a7ehuo) [inline runtime/compiler/ilgen/J9ByteCodeIlGenerator.hpp:208]: Just a very minor comment: `narrowIntStore` can be invoked if the type is `TR::Int32` and no action will be taken. I wonder if `narrowIntStore` should be updated to something like `narrowIntStoreIfRequired` to reflect that not all integral type are going through narrowing conversion.

**Review by a7ehuo — COMMENTED** [review]
Thank you very much for making this change @BradleyWood! 

LGTM overall. I have just two very minor comments. Could the commit message be updated to include the issue number #16498?

**Comment by BradleyWood** [comment]
See the [force-push](https://github.com/eclipse-openj9/openj9/compare/0c89f2941b6d23780dd0ebd6157c73042b52e5ed..0426748717b67aac72ee68d5afb5fba143d3b8db)

**Comment by a7ehuo** [comment]
@BradleyWood Looks like the latest force-push 0426748 might be based off an older version of this branch. The following code is removed in 0426748 which I think should be included to address Henry's previous comments 

Screenshot 2024-03-05 at 4 21 28 PM
Screenshot 2024-03-05 at 4 21 36 PM

**Comment by BradleyWood** [comment]
Sorry, thanks for catching that

**Comment by BradleyWood** [comment]
Fixed. @hzongaro @a7ehuo

**Review by a7ehuo — APPROVED** [review]
Thank you for the update Brad! LGTM

**Review by hzongaro — APPROVED** [review]
Looks good.  Thanks!

**Comment by hzongaro** [comment]
Jenkins test sanity all jdk8,jdk11,jdk17,jdk21

**Comment by hzongaro** [comment]
[zLinux JDK 17](https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_sanity.functional_s390x_linux_Personal/439/) and [aarch64 JDK 21](https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_sanity.functional_aarch64_linux_Personal/125/) test failures appears to be known issue #18527 (or a [variant](../issues/18527#issuecomment-1959521554)).

[aarch64 JDK 17](https://openj9-jenkins.osuosl.org/job/Build_JDK17_aarch64_linux_Personal/713/) and [AIX JDK 21](https://openj9-jenkins.osuosl.org/job/Build_JDK21_ppc64_aix_Personal/130/) build failures appear to be infrastructure-related.  Rerunning those two.

**Comment by hzongaro** [comment]
Jenkins test sanity aix jdk21

**Comment by hzongaro** [comment]
Testing was successful, aside from the known failures noted [above](../18904#issuecomment-1982318887).  Merging.

**Comment by hzongaro** [comment]
@BradleyWood, may I ask you to open a pull request to get this change delivered to the 0.44 release branch?

**Comment by gacholio** [comment]
AFIAK, my changes have not been backported to 0.44 (unless the split was after my commits).

**Comment by pshipton** [comment]
https://github.com/eclipse-openj9/openj9/pull/18929 is included in 0.44, the split was done after it was merged.

**Pull request merged**